app/models/mixins/scanning_mixin.rb

```diff
@@ -0,0 +1,377 @@
+      rescue REXML::ParseException
+        n = MiqXml.load(xml_file, :nokogiri)
+        xml_file = n.to_s
+        doc = MiqXml.load(xml_file)
+      end
+
+      taskid = doc.root.attributes["taskid"]
+      _log.info("TaskId = [#{taskid}]")
+      unless taskid.blank?
+        name =  begin
+                  File.basename(doc.root.elements[1].elements[1].attributes["original_filename"], ".*")
+                rescue
+                  "vmscan"
+                end
+        job = Job.find_by_guid(taskid)
+        raise "Unable to process data for job with id <#{taskid}>.  Job not found." if job.nil?
+        begin
+          job.signal(:data, xml_file)
+        rescue => err
+          _log.error("Processing xml for [#{name}] [#{err}]")
+          _log.error("Processing xml for [#{name}] #{err.backtrace.join("\n")}")
+        end
+      end
+
+      case doc.root.name.downcase
+      when "summary"
+        _log.info("Summary XML received. [#{doc.root.to_s[0..100]}]")
+      when "vmmetadata" # TODO: - should be "metadata"
+        # doc being sent up has a some extra header elements we need to remove (<vmmetadata>, and <item>)
+        begin
+          # Reset the root of the xml document to match the expected starting point
+          doc.root = doc.root.elements[1].elements[1]
+        rescue => err
+          _log.error "Invalid xml error [#{err}] for xml:[#{doc}]"
+        end
+        target.add_elements(doc)
+        target.save!
+      when "vmevents" # TODO: - should be "events"
+        target.add_elements(doc)
+        target.save!
+      else
+        target.add_elements(doc)
+        target.save!
+      end
+    end
+  end # ClassMethods
+
+  # Process XML documents from VM scans
+  def add_elements(xml_node)
+    return if xml_node.nil?
+    _log.info("Adding XML elements for [#{id}] from [#{xml_node.root.name}]")
+    updated = false
+
+    # Find out what XML file document we are being passed.
+    case xml_node.root.name
+    when "miq"
+      for element_class in [OperatingSystem, Account, SystemService, GuestApplication, Patch, Network]
+        begin
+          element_class.add_elements(self, xml_node)
+          updated = true
+        rescue Exception => err
+          _log.log_backtrace(err)
+        end
+      end
+    when "scan_profiles"
+      ScanItem.add_elements(self, xml_node)
+      updated = true
+    when "vm_configuration" # TODO: should be "configuration"?
+      Hardware.add_elements(self, xml_node)
+      updated = true
+    when "vmevents" # TODO: should be "events"?
+      # Record vm operational and configuration events
+      MiqEvent.add_elements(self, xml_node)
+    end
+    # Update the last sync time if we did something
+    # self.last_sync_on = Time.new.utc  if updated == true
+    self.last_sync_on = Time.at(xml_node.root.attributes["created_on"].to_i).utc if updated == true && xml_node.root.attributes["created_on"]
+    save
+    hardware.save unless hardware.nil?
```

Comment:
LOL.
I interpreted `and not` as English instead of Ruby. So, I thought you meant `if self.respond_to?(:hardware)` instead of `unless hardware.nil?`. Rubocop would complain about using `and not`.
So: `hardware.save if self.respond_to?(:hardware) && !hardware.nil?` it is.